examples/rlhf/train_rlhf.py

```diff
@@ -0,0 +1,385 @@
+        (0, max_new_tokens - scores.shape[1]),
+        value=0,
+    )
+
+    # get the scores and normalise for log probabilities
+    attention_mask = (generated != EOS_TOKEN_ID).to(torch.int64)
+    logits = model(
+        input_ids=generated, attention_mask=attention_mask, return_dict=True
+    ).logits
+    logprobs = logprobs_of_labels(logits[:, :-1], generated[:, 1:])
+    ref_logits = ref_model(
+        input_ids=generated.to(ref_model.device),
+        attention_mask=attention_mask.to(ref_model.device),
+        return_dict=True,
+    ).logits.to(logits.device)
+    ref_logprobs = logprobs_of_labels(ref_logits[:, :-1], generated[:, 1:])
+    log_ratio = (logprobs - ref_logprobs) * attention_mask[:, :-1]
+    log_ratio = torch.stack(
+        [
+            row[rindex - 1 : rindex + max_new_tokens - 1]
+            for row, rindex in zip(log_ratio, batch.prompt_rindex)
+        ],
+        dim=0,
+    )
+    return generated, log_probs_gen, log_ratio
+
+
+@torch.no_grad()
+def create_rollout_td(
+    batch, generated, reward_model, log_probs, log_ratio, max_new_tokens=50, kl_coef=0.1
+):
+    """
+    This function takes a batch plus the generated tokens and replicates the tensordict
+    structure that would have been obtained from a rollout with a TorchRL env that
+    sampled one token each timestep.
+    """
+    rollout_generated = torch.stack(
+        [
+            torch.stack(
+                [
+                    torch.where(
+                        torch.arange(row.shape[0], device=generated.device)
+                        < rindex + i,
+                        row,
+                        EOS_TOKEN_ID,
+                    )
+                    # + 1 so that we get prompt and full generated sequence as first
+                    # and last row respectively
+                    for i in range(max_new_tokens + 1)
+                ]
+            )
+            for rindex, row in zip(batch.prompt_rindex, generated)
+        ],
+    )
+    rollout_attention_mask = (rollout_generated != EOS_TOKEN_ID).to(torch.int64)
+
+    # done is True when we either first sample an EOS token or reach the maximum number
+    # of generated tokens
+    done_idx = torch.minimum(
+        (generated != EOS_TOKEN_ID).sum(dim=-1) - batch.prompt_rindex,
+        torch.tensor(max_new_tokens),
+    )
+    done = (
+        torch.arange(max_new_tokens, device=generated.device) == done_idx[:, None]
+    ).unsqueeze(-1)
+
+    # the sequence of actions for each trajectory is just the generated token ids
+    action_idx = torch.stack(
+        [
+            torch.arange(i, i + max_new_tokens, device=generated.device)
+            for i in batch.prompt_rindex
+        ]
+    )
+    action = generated[
+        torch.arange(generated.shape[0], device=generated.device)[:, None],
+        action_idx,
+    ]
+
+    # calculate the reward for the finished sequence
+    _, end_scores = reward_model(
```

Comment:
is this a tensordict module?
Are we using dispatch?
If so I would either not use it with dispatch but with a plain tensordict or use select_out_keys